### DISCUSSION TITLE
how to make bear work with different compilers

### DISCUSSION THREAD

**Discussion by jpy139** [rizsotto/Bear#534] [Q&A]
I'm trying to setup a development environment for a company project. 
After installing bear, I found it works with the gcc I installed via apt, but when used together with the cross compiler that I installed manually, I always got the following error:

`ERROR: ld.so: object '/usr/local/$LIB/bear/libexec.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.`

I thought it might be because the cross compiler is 32bit and the one that works fine is 64bit, but this problem still exist after I installed both 32bit and 64bit bear accroding to wiki. I also checked the previously reported issues but didn't get a working solution. So I wonder if you have time to help analyze.

Environment:

OS name: Linux

OS version: Ubuntu 22.04.2 LTS

OS architecture: x86_64

Bear version: 3.1.2

Bear install method: built from source

What architecture you are trying to compile for?  arm 32bit

cross compiler：arm-none-eabi-gcc
`Using built-in specs.
COLLECT_GCC=arm-none-eabi-gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/gcc-arm-none-eabi-5_4-2016q3/bin/../lib/gcc/arm-none-eabi/5.4.1/lto-wrapper
Target: arm-none-eabi
Configured with: /home/build/work/GCC-5-build/src/gcc/configure --target=arm-none-eabi --prefix=/home/build/work/GCC-5-build/install-native --libexecdir=/home/build/work/GCC-5-build/install-native/lib --infodir=/home/build/work/GCC-5-build/install-native/share/doc/gcc-arm-none-eabi/info --mandir=/home/build/work/GCC-5-build/install-native/share/doc/gcc-arm-none-eabi/man --htmldir=/home/build/work/GCC-5-build/install-native/share/doc/gcc-arm-none-eabi/html --pdfdir=/home/build/work/GCC-5-build/install-native/share/doc/gcc-arm-none-eabi/pdf --enable-languages=c,c++ --enable-plugins --disable-decimal-float --disable-libffi --disable-libgomp --disable-libmudflap --disable-libquadmath --disable-libssp --disable-libstdcxx-pch --disable-nls --disable-shared --disable-threads --disable-tls --with-gnu-as --with-gnu-ld --with-newlib --with-headers=yes --with-python-dir=share/gcc-arm-none-eabi --with-sysroot=/home/build/work/GCC-5-build/install-native/arm-none-eabi --build=i686-linux-gnu --host=i686-linux-gnu --with-gmp=/home/build/work/GCC-5-build/build-native/host-libs/usr --with-mpfr=/home/build/work/GCC-5-build/build-native/host-libs/usr --with-mpc=/home/build/work/GCC-5-build/build-native/host-libs/usr --with-isl=/home/build/work/GCC-5-build/build-native/host-libs/usr --with-cloog=/home/build/work/GCC-5-build/build-native/host-libs/usr --with-libelf=/home/build/work/GCC-5-build/build-native/host-libs/usr --with-host-libstdcxx='-static-libgcc -Wl,-Bstatic,-lstdc++,-Bdynamic -lm' --with-pkgversion='GNU Tools for ARM Embedded Processors' --with-multilib-list=armv6-m,armv7-m,armv7e-m,armv7-r,armv8-m.base,armv8-m.main
Thread model: single
gcc version 5.4.1 20160919 (release) [ARM/embedded-5-branch revision 240496] (GNU Tools for ARM Embedded Processors) `

working compiler：gcc 11.3.0
`Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/11/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:amdgcn-amdhsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 11.3.0-1ubuntu1~22.04.1' --with-bugurl=file:///usr/share/doc/gcc-11/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-11 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --enable-cet --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-11-aYxV0E/gcc-11-11.3.0/debian/tmp-nvptx/usr,amdgcn-amdhsa=/build/gcc-11-aYxV0E/gcc-11-11.3.0/debian/tmp-gcn/usr --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu --with-build-config=bootstrap-lto-lean --enable-link-serialization=2
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.3.0 (Ubuntu 11.3.0-1ubuntu1~22.04.1)`

And some other information of my environment:
`$ bear --version
bear 3.1.2`

`$ ll /usr/local/lib/x86_64-linux-gnu/bear/
total 752
drwxrwxr-x 3 root root   4096 Jun 13 16:53 ./
drwxrwxr-x 3 root root   4096 Jun 13 17:38 ../
-rw-r--r-- 1 root root  37976 Jun 13 16:48 libexec.so
-rwxr-xr-x 1 root root 715368 Jun 13 16:51 wrapper*
drwxr-xr-x 2 root root   4096 Jun 13 16:53 wrapper.d/`

`$ ll /usr/local/lib/i386-linux-gnu/bear/
total 752
drwxrwxr-x 3 root root   4096 Jun 13 16:45 ./
drwxrwxr-x 3 root root   4096 Jun 13 16:45 ../
-rw-r--r-- 1 root root  37976 Jun 13 16:42 libexec.so
-rwxr-xr-x 1 root root 715368 Jun 13 16:45 wrapper*
drwxr-xr-x 2 root root   4096 Jun 13 16:45 wrapper.d/`

**Comment by rizsotto** [comment]
@jpy139, your set up looks good. The multilib setup is not tested by CI, and it might be broken. Sorry for that.

**Reply by jpy139** [reply]
@rizsotto Hello, and thank you for your reply, 
Currently I just want to get Bear+arm-none-eabi-gcc working, can you let me know how should I fix it next?
